src/main/java/net/openhft/chronicle/core/threads/ThreadLock.java

```diff
@@ -0,0 +1,134 @@
+            if (tryLock(threadId))
+                return;
+            Jvm.nanoPause();
+        }
+        if (busyLockSlower(threadId))
+            return;
+        forceUnlockAndRetry(threadId);
+    }
+
+    @SuppressWarnings("java:S3776")
+    private boolean busyLockSlower(int threadId) throws InterruptedRuntimeException {
+        long endMS = System.currentTimeMillis() + timeoutMs;
+        do {
+            for (int i = 0; i < busyLockSlowerCount; i++) {
+                if (tryLock(threadId))
+                    return true;
+                Thread.yield();
+            }
+            if (METRICS.supportsProc) {
+                int lockedThread = (int) twoThreadId.getVolatileValue();
+                if (lockedThread == 0)
+                    continue;
+                if (!isThreadRunning(lockedThread)) {
+                    Jvm.warn().on(getClass(), "ThreadId " + lockedThread + " died while holding a lock");
+                    return false;
+                }
+            }
+            if (Thread.currentThread().isInterrupted())
+                throw new InterruptedRuntimeException();
+
+        } while (System.currentTimeMillis() < endMS);
+        return false;
+    }
+
+    private boolean isThreadRunning(int lockedThread) {
+        return new File("/proc/" + lockedThread).exists();
+    }
+
+    @SuppressWarnings("java:S3358")
+    private void forceUnlockAndRetry(int threadId) {
+        long twoThreadId0 = twoThreadId.getVolatileValue();
+        int threadId0 = (int) twoThreadId0;
+
+        // unlock the previous thread
+        if (twoThreadId.compareAndSwapValue(twoThreadId0, ((long) threadId0 << 32) | UNLOCKED)) {
+            String status = METRICS.supportsProc
+                    ? isThreadRunning(threadId0) ? "running" : "dead"
+                    : "unknown";
+            Jvm.warn().on(getClass(), "Successfully forced an unlock for threadId: " + threadId + ", previous thread held by: " + threadId0 + ", status: " + status);
+        } else {
+            Jvm.warn().on(getClass(), "Failed to forced an unlock for threadId: " + threadId);
+        }
+        lock(threadId);
+    }
+
+    public void unlock(int threadId) throws IllegalStateException {
+        long twoThreadId0 = twoThreadId.getVolatileValue();
+        int threadId0 = (int) twoThreadId0;
+        if (threadId0 != threadId) {
+            unlock2(threadId, twoThreadId0, threadId0);
+            return;
+        }
+
+        if (twoThreadId.compareAndSwapValue(twoThreadId0, ((long) threadId << 32) | UNLOCKED))
+            return;
+        Jvm.warn().on(getClass(), "Failed to unlock");
+    }
+
+    private void unlock2(int threadId, long twoThreadId0, int threadId0) {
+        if (threadId0 == 0) {
+            int prevThreadId = (int) (twoThreadId0 >>> 32);
+            if (prevThreadId == threadId)
+                throw new IllegalStateException("Lock already unlocked by threadId " + threadId);
```

Comment:
Also, shouldn't unlock be idempotent? maybe log a warning? throwing seems extreme here?

Reply:
It suggests a coding error to attempt to unlock multiple times. You have the option to catch the exception if you are just doing it as you are not sure if you have already unlocked it.